batch/batch/front_end/front_end.py

```diff
@@ -1448,46 +1733,58 @@ async def update(tx: Transaction):
             """
 SELECT job_groups_cancelled.id IS NOT NULL AS cancelled
 FROM batches
-LEFT JOIN job_groups_cancelled ON batches.id = job_groups_cancelled.id
-WHERE batches.id = %s AND user = %s AND NOT deleted
+LEFT JOIN job_groups_cancelled ON batches.id = job_groups_cancelled.id AND job_groups_cancelled.job_group_id = %s
+WHERE batches.id = %s AND batches.user = %s AND NOT deleted
 FOR UPDATE;
 """,
-            (batch_id, user),
+            (ROOT_JOB_GROUP_ID, batch_id, user),
         )
         if not record:
             raise web.HTTPNotFound()
         if record['cancelled']:
-            raise web.HTTPBadRequest(reason='Cannot submit new jobs to a cancelled batch')
+            raise web.HTTPBadRequest(reason='Cannot submit new jobs or job groups to a cancelled batch')
 
         now = time_msecs()
 
         record = await tx.execute_and_fetchone(
             """
-SELECT update_id, start_job_id, n_jobs FROM batch_updates
+SELECT update_id, start_job_id, n_jobs, start_job_group_id, n_job_groups FROM batch_updates
 WHERE batch_id = %s
 ORDER BY update_id DESC
 LIMIT 1;
```

Comment:
Hmm. Now I wonder why we don't need a FOR UPDATE here too to ensure that two racing updates don't both get the same start_job_id.

Reply:
It should be locked in the same transaction. See the comment in the code in the original lines 1444-1445.